### ISSUE TITLE
[DataGrid] column reordering glitches.

### ISSUE THREAD

**Issue by yaredtsy** [mui/mui-x#8316]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Steps to reproduce 🕹

Link to live example: https://codesandbox.io/s/jovial-leavitt-y22uls

Steps:
1. try to reorder the column username and id

### Current behavior 😯

it keeps reordering continuously.

https://user-images.githubusercontent.com/34584266/226616671-26a02fa6-c074-48b1-a5bc-14a392aa2a87.mp4

### Expected behavior 🤔

it should not reorder continuously. 

### Your environment 🌎

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Comment by yaredtsy** [comment]
I have noticed that the same problem occurs when reordering rows, and I have created a pull request for it at #7006. The same logic could also be applied to this issue.

**Comment by cherniavskii** [comment]
This is a regression that was released in [6.0.0-beta.3](https://github.com/mui/mui-x/releases/tag/v6.0.0-beta.3) 
Demo with 6.0.0-beta.2: https://codesandbox.io/s/nifty-lewin-u1d917?file=/package.json
Demo with 6.0.0-beta.3: https://codesandbox.io/s/eager-buck-pf4qex?file=/package.json

**Comment by cherniavskii** [comment]
The regression seems to come from https://github.com/mui/mui-x/pull/6580
This is the first PR where the issue is reproducible - see https://deploy-preview-6580--material-ui-x.netlify.app/x/react-data-grid/column-ordering/
